taichi/backends/cuda/cuda_driver.cpp

```diff
@@ -0,0 +1,50 @@
+#include "cuda_driver.h"
+
+#include "taichi/system/dynamic_loader.h"
+#include "taichi/backends/cuda/cuda_context.h"
+
+TLANG_NAMESPACE_BEGIN
+
+std::string get_cuda_error_message(uint32 err) {
+  const char *err_name_ptr;
+  const char *err_string_ptr;
+  CUDADriver::get_instance_without_context().get_error_name(err, &err_name_ptr);
+  CUDADriver::get_instance_without_context().get_error_string(err,
+                                                              &err_string_ptr);
+  return fmt::format("CUDA Error {}: {}", err_name_ptr, err_string_ptr);
+}
+
+std::unique_ptr<CUDADriver> CUDADriver::instance;
+
+CUDADriver::CUDADriver() {
+#if defined(TI_PLATFORM_LINUX)
+  loader = std::make_unique<DynamicLoader>("libcuda.so");
+#else
+  loader = std::make_unique<DynamicLoader>("nvcuda.dll");
+#endif
+
+  loader->load_function("cuGetErrorName", get_error_name);
+  loader->load_function("cuGetErrorString", get_error_string);
+
+#define PER_CUDA_FUNCTION(name, symbol_name, ...) \
+  name.set(loader->load_function(#symbol_name));  \
+  name.set_names(#name, #symbol_name);
+#include "taichi/backends/cuda/cuda_driver_functions.inc.h"
+#undef PER_CUDA_FUNCTION
+
+  TI_INFO("CUDA driver loaded");
+}
+
+// This is for initializing the CUDA context itself
+CUDADriver &CUDADriver::get_instance_without_context() {
+  if (!instance)
```

Comment:
C++'s singleton pattern is notoriously hard to be implemented correctly, especially when multi-threading comes into play (which is why i pushed back so hard when I see global variables abused). At a minimum, this requires [double checked locking pattern (DCLP)](https://preshing.com/20130930/double-checked-locking-is-fixed-in-cpp11/). But even DCLP is not enough if memory fences are not properly installed. 

Fortunately, since C++11, the compiler guarantees a thread-safe initialization of the  local static variables (https://stackoverflow.com/a/8102145/12003165). So I think we can transform this into:

```cpp
CUDADriver &CUDADriver::get_instance_without_context() {
  static CUDADriver* instance = new CUDADriver();  // Thread safety guaranteed by C++ compiler
  return *instance;
}
```

Note the intentional usage of raw pointer. Yes, its memory will be leaked. But that's OK because the process will clean it up anyway. The main reason for this is that, global variable's destruction order is nondeterministic, and it's possible that while the variable is destructed on one thread, another thread still tries to use it. So we forbid the destructor from running to begin with (After all, raw pointers do not have destructors). See https://stackoverflow.com/a/13047943/12003165

---

At last, all the discussion here is about the thread safety of initializing the local static variable. It says nothing about the thread safety of `CUDADriver` itself. So if its members can be accessed in multi-threads, i think we still need a `std::mutex` to guard them.. 
FYI, [clang thread safety analysis](https://clang.llvm.org/docs/ThreadSafetyAnalysis.html) is a super cool annotation system. It checks for possible data races at compilation time! We might want to integrate this in the future..

Reply:
Thank you so much! This is fixed in the latest commit.